catalog/app/components/Assistant/Model/GlobalContext/preview.ts

```diff
@@ -0,0 +1,439 @@
+    getObject(handle: S3ObjectLocation): AWSEffect<AWSSDK.S3.GetObjectOutput>
+  }
+>() {}
+
+export const fromS3Client = (client: AWSSDK.S3) =>
+  Eff.Layer.succeed(S3, {
+    headObject: (handle) =>
+      Eff.Effect.tryPromise({
+        try: () =>
+          client
+            .headObject({
+              Bucket: handle.bucket,
+              Key: handle.key,
+              VersionId: handle.version,
+            })
+            .promise(),
+        catch: (e) => e as AWSSDK.AWSError,
+      }),
+    getObject: (handle) =>
+      Eff.Effect.tryPromise({
+        try: () =>
+          client
+            .getObject({
+              Bucket: handle.bucket,
+              Key: handle.key,
+              VersionId: handle.version,
+            })
+            .promise(),
+        catch: (e) => e as AWSSDK.AWSError,
+      }),
+  })
+
+export interface S3SignerOptions {
+  urlExpiration?: number // in seconds
+  forceProxy?: boolean
+}
+
+export class S3Signer extends Eff.Context.Tag('S3Signer')<
+  S3Signer,
+  {
+    sign(handle: S3ObjectLocation, options?: S3SignerOptions): Eff.Effect.Effect<string>
+  }
+>() {}
+
+export const fromS3Signer = (
+  signer: (handle: S3ObjectLocation, options?: S3SignerOptions) => string,
+) =>
+  Eff.Layer.succeed(S3Signer, {
+    sign: (...args) => Eff.Effect.sync(() => signer(...args)),
+  })
+
+// The document file name can only contain:
+// - alphanumeric characters
+// - whitespace characters
+// - hyphens
+// - parentheses and square brackets
+// The name can't contain more than one consecutive whitespace character
+const normalizeDocumentName = (name: string) =>
+  name
+    .replace(/[^a-zA-Z0-9\s\-\(\)\[\]]/g, ' ') // Remove invalid characters
+    .replace(/\s+/g, ' ') // Replace multiple whitespace characters with a single space
+    .trim() // Remove leading and trailing whitespace
+
+const THRESHOLD = 500 * 1024 // 500 KiB
+
+const GetObjectSchema = S.Struct({
+  bucket: S.String,
+  key: S.String,
+  version: S.optional(S.String).annotations({}),
+  // XXX: force type?
+}).annotations({
+  description: 'Get contents and metadata of an S3 object',
+})
+
+// return format:
+// - metadata block (text or json)
+// - content block (json | text | image | document)
+export function useGetObject() {
+  const s3Client = AWS.S3.use()
+  const s3Signer = AWS.Signer.useS3Signer()
+
+  return Tool.useMakeTool(
+    GetObjectSchema,
+    Eff.flow(
+      getObject,
+      Eff.Effect.map(Eff.Option.some),
+      Eff.Effect.provide(fromS3Client(s3Client)),
+      Eff.Effect.provide(fromS3Signer(s3Signer)),
+    ),
+    [s3Client, s3Signer],
+  )
+}
+
+const getObject = (handle: S3ObjectLocation) =>
+  Log.scoped({
+    name: `${MODULE}.getObject`,
+    enter: [Log.br, 'handle:', handle],
+  })(
+    Eff.Effect.gen(function* () {
+      const s3 = yield* S3
+      const headE = yield* Eff.Effect.either(s3.headObject(handle))
+      if (Eff.Either.isLeft(headE)) {
```

Comment:
> As I understand, left is an error, and right is success.

yep

> Are you going to make it more readable by creating some kind of custom Either

not sure what you mean here

> or did you already do this with Tool.fail/succeed?

well, tool result type is a little bit special, since it's not a success/error in pure sense

Reply:
> Are you going to make it more readable by creating some kind of custom Either

I would like to use `Response.error`/`Response.payload` or something similar, instead of keeping in mind that `Either.left` is an error and `Either.right` is a response payload.